File: app/src/main/java/ru/mpei/vmss/myapplication/adapters/TasksAdapter.java
package ru.mpei.vmss.myapplication.adapters;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.TextView;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.List;

import ru.mpei.vmss.myapplication.R;
import ru.mpei.vmss.myapplication.pojo.Task;

import static ru.mpei.vmss.myapplication.fragments.User.isAuth;
import static ru.mpei.vmss.myapplication.fragments.User.userId;

public class TasksAdapter extends BaseExpandableListAdapter {

    private List<Task> elements;
    private LayoutInflater inflater;
    private Context context;

    public TasksAdapter(Context context, List<Task> elements, int type) {
        this.elements = elements;
        inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        this.context = context;
    }

    @Override
    public int getGroupCount() {
        return elements.size();
    }

    @Override
    public int getChildrenCount(int groupPosition) {
        return 1;
    }

    @Override
    public Object getGroup(int groupPosition) {
        return elements.get(groupPosition);
    }

    @Override
    public Object getChild(int groupPosition, int childPosition) {
        return null;
    }

    @Override
    public long getGroupId(int groupPosition) {
        return groupPosition;
    }

    @Override
    public long getChildId(int groupPosition, int childPosition) {
        return 0;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @SuppressLint("SetTextI18n")
    @Override
    public View getGroupView(int groupPosition, boolean isExpanded, View convertView, ViewGroup parent) {
        View view = convertView;
        if(view == null){
            view = inflater.inflate(R.layout.tasks_element, parent, false);
        }

        TextView taskNameText = view.findViewById(R.id.taskNameText);
        taskNameText.setText(getTasksElement(groupPosition).getTaskName());

        TextView taskPriceText = view.findViewById(R.id.taskPriceText);
        taskPriceText.setText(getTasksElement(groupPosition).getPrice() + " б.");

        TextView taskLocationText = view.findViewById(R.id.taskLocationText);
        taskLocationText.setText("Место: " + getTasksElement(groupPosition).getLocation());

        return view;
    }

    private Task getTasksElement(int position) {
        return (Task) getGroup(position);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public View getChildView(int groupPosition, int childPosition, boolean isLastChild, View convertView, ViewGroup parent) {
        View view = convertView;

        if(view == null){
            view = inflater.inflate(R.layout.available_tasks_element_child, parent, false);
        }

        TextView availableDescription = view.findViewById(R.id.availableDescriptionText);
        availableDescription.setText(getTasksElement(groupPosition).getTaskDescription());

        TextView availableBeginTime = view.findViewById(R.id.availableBeginText);
        availableBeginTime.setText("Начало: " + getTasksElement(groupPosition).getStartDate());

        TextView availableEndTime = view.findViewById(R.id.availableEndText);
        availableEndTime.setText("Конец: " + getTasksElement(groupPosition).getEndDate());

        TextView availableRefuseBefore = view.findViewById(R.id.availableCommentText);
        availableRefuseBefore.setText("От задания можно отказаться не позднее, чем: " + getTasksElement(groupPosition).getRefuse_info());

        view.findViewById(R.id.availableAgreeButton).setOnClickListener(v -> {
            if(isAuth) {
                try {
                    JSONObject body = new JSONObject();
                    body.put("user_id", userId);
                    body.put("task_id", getTasksElement(groupPosition).getID());
                    JsonObjectRequest request = new JsonObjectRequest(Request.Method.POST, context.getString(R.string.takeTaskUrl), body,
                            response -> {
                            },
                            error -> {
                            });

                    RequestQueue requestQueue = Volley.newRequestQueue(context);
                    requestQueue.add(request);

                    elements.remove(groupPosition);
                    notifyDataSetChanged();
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            } else {
                Toast.makeText(context, "Необходимо авторизоваться", Toast.LENGTH_SHORT).show();
            }
        });


       return view;
    }

    @Override
    public boolean isChildSelectable(int groupPosition, int childPosition) {
        return false;
    }

}

File: app/src/main/java/ru/mpei/vmss/myapplication/adapters/TasksAdapter.kt
package ru.mpei.vmss.myapplication.adapters

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import android.widget.TextView
import android.widget.Toast
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import org.json.JSONException
import org.json.JSONObject
import ru.mpei.vmss.myapplication.R
import ru.mpei.vmss.myapplication.fragments.User
import ru.mpei.vmss.myapplication.pojo.Task

class TasksAdapter(private val context: Context, private val elements: MutableList<Task>) : BaseExpandableListAdapter() {
    private val inflater: LayoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

    init{
    }

    override fun getGroupCount(): Int {
        return elements.size
    }

    override fun getChildrenCount(groupPosition: Int): Int {
        return 1
    }

    override fun getGroup(groupPosition: Int): Any {
        return elements[groupPosition]
    }

    override fun getChild(groupPosition: Int, childPosition: Int): Any? {
        return null
    }

    override fun getGroupId(groupPosition: Int): Long {
        return groupPosition.toLong()
    }

    override fun getChildId(groupPosition: Int, childPosition: Int): Long {
        return 0
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    @SuppressLint("SetTextI18n")
    override fun getGroupView(groupPosition: Int, isExpanded: Boolean, convertView: View, parent: ViewGroup): View {
        var view = convertView
        if (view == null) {
            view = inflater.inflate(R.layout.tasks_element, parent, false)
        }
        val taskNameText = view.findViewById<TextView>(R.id.taskNameText)
        taskNameText.text = getTasksElement(groupPosition).taskName
        val taskPriceText = view.findViewById<TextView>(R.id.taskPriceText)
        taskPriceText.text = getTasksElement(groupPosition).price + " б."
        val taskLocationText = view.findViewById<TextView>(R.id.taskLocationText)
        taskLocationText.text = "Место: " + getTasksElement(groupPosition).location
        return view
    }

    private fun getTasksElement(position: Int): Task {
        return getGroup(position) as Task
    }

    @SuppressLint("SetTextI18n")
    override fun getChildView(groupPosition: Int, childPosition: Int, isLastChild: Boolean, convertView: View, parent: ViewGroup): View {
        var view = convertView
        if (view == null) {
            view = inflater.inflate(R.layout.available_tasks_element_child, parent, false)
        }
        val availableDescription = view.findViewById<TextView>(R.id.availableDescriptionText)
        availableDescription.text = getTasksElement(groupPosition).taskDescription
        val availableBeginTime = view.findViewById<TextView>(R.id.availableBeginText)
        availableBeginTime.text = "Начало: " + getTasksElement(groupPosition).startDate
        val availableEndTime = view.findViewById<TextView>(R.id.availableEndText)
        availableEndTime.text = "Конец: " + getTasksElement(groupPosition).endDate
        val availableRefuseBefore = view.findViewById<TextView>(R.id.availableCommentText)
        availableRefuseBefore.text = "От задания можно отказаться не позднее, чем: " + getTasksElement(groupPosition).refuseInfo
        view.findViewById<View>(R.id.availableAgreeButton).setOnClickListener { v: View? ->
            if (User.isAuth) {
                try {
                    val body = JSONObject()
                    body.put("user_id", User.userId)
                    body.put("task_id", getTasksElement(groupPosition).id)
                    val request = JsonObjectRequest(Request.Method.POST, context.getString(R.string.takeTaskUrl), body,
                            Response.Listener { },
                            Response.ErrorListener { })
                    val requestQueue = Volley.newRequestQueue(context)
                    requestQueue.add(request)
                    elements.removeAt(groupPosition)
                    notifyDataSetChanged()
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            } else {
                Toast.makeText(context, "Необходимо авторизоваться", Toast.LENGTH_SHORT).show()
            }
        }
        return view
    }

    override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
        return false
    }

}